Clarify v2 stream specification

sgallagher opened the pull request:
Related to [BZ 1799036](https://bugzilla.redhat.com/show_bug.cgi?id=1799036)

Several fields (notably those that provide `N:S:V:C:A`) were marked as being "optional" when that was really only because the packager should not be setting them. 

Also separates the different spec sections with blank lines for readability.

Signed-off-by: Stephen Gallagher <sgallagh@redhat.com>

Conan-Kudo commented on yaml_specs/modulemd_packager_v2.yaml:
```diff
@@ -36,8 +39,6 @@ data:
     # against.  Multiple builds result in multiple output modulemd files.
     # See below for an example.
-    # TODO: Provides, conflicts, obsoletes, recommends, etc.
-    #       Do we even need those?
```
We are probably going to need these...

Conan-Kudo commented on yaml_specs/modulemd_stream_v2.yaml:
```diff
@@ -91,8 +126,6 @@ data:
     # against.  Multiple builds result in multiple output modulemd files.
     # See below for an example.
-    # TODO: Provides, conflicts, obsoletes, recommends, etc.
-    #       Do we even need those?
```
As said above, this will likely become necessary soon. Not sure why we don't have it yet, actually...

sgallagher replied:
I removed these from the specification because I didn't think *proposed* concepts belong in the spec (it confuses the reader). Those should be filed as tickets in the issue tracker instead.

Conan-Kudo replied:
If you move them to the issue tracker, then this will look good to merge.

sgallagher replied:
https://github.com/fedora-modularity/libmodulemd/issues/448

mmathesius commented on yaml_specs/modulemd_packager_v2.yaml:
```diff
@@ -3,15 +3,18 @@
     description: >-
         A module for the demonstration of the metadata format. Also,
         the obligatory lorem ipsum dolor sit amet goes right here.
+
```
While adding blank lines for clarity, add one between `license` and `dependencies` below.

mmathesius commented on yaml_specs/modulemd_stream_v2.yaml:
```diff
@@ -4,40 +4,66 @@ document: modulemd
+    #
+    # When included as input to the build system, the behavior is unspecified.
+    #
+    # Mandatory for module metadata in a yum/dnf repository.
```
I think this needs something noted about what the value `0` means. Since `version` is an integer, I think `0` is not allowed as a module version identifier, and gets treated treated the same as if it were unspecified.

sgallagher replied:
It's actually the other way around. Unspecified is the same as `version==0` which is a valid version. The special-case here is that if you input `0`, you won't see it in the output. I'll try to clear that up. I guess that means this field isn't *actually* mandatory, though.

mmathesius commented on yaml_specs/modulemd_stream_v2.yaml:
```diff
@@ -4,40 +4,66 @@ document: modulemd
     # Filled in by the buildsystem.  A short hash of the module's name,
-    # stream, version and its expanded runtime dependencies.
+    # stream, version and its expanded runtime dependencies. The exact
+    # mechanism of generating the has is unspecified except that it must be
```
```suggestion
    # mechanism for generating the hash is unspecified except that it must be
```

goosemania commented:
:+1:  Thanks for the clarification, @sgallagher !

mmathesius commented on yaml_specs/modulemd_stream_v2.yaml:
```diff
@@ -4,40 +4,66 @@ document: modulemd
+    # stream, version and its expanded runtime dependencies. The exact
+    # mechanism of generating the has is unspecified except that it must be
+    # stable.
+    # When included as input to the build system, the behavior is unspecified.
```
Please empty comment lines above and below for clarity as was done for `name`, `stream`, and `version`.

mmathesius commented on yaml_specs/modulemd_stream_v2.yaml:
```diff
@@ -150,198 +264,291 @@ data:
                 # Use of both buildafter and buildorder in the same document is
                 # prohibited, as they will conflict.
+                #
+                #
```
Remove extraneous empty line.

mmathesius commented on yaml_specs/modulemd_stream_v2.yaml:
```diff
@@ -350,29 +557,49 @@ data:
+                # Type: MANDATORY
                 rationale: Included in the stack, just because.
+
                 # Link to VCS repository that contains the modulemd file
```
Add `# repository:` line here.

mmathesius commented on yaml_specs/modulemd_stream_v2.yaml:
```diff
@@ -350,29 +557,49 @@ data:
+                # Type: AUTOMATIC
                 ref: somecoolbranchname
+
+                # buildorder
```
```suggestion
                # buildorder:
```

mmathesius commented on yaml_specs/modulemd_stream_v2.yaml:
```diff
@@ -381,24 +608,50 @@ data:
             - xxx-0:1-1.module_deadbeef.i686
             - xyz-0:1-1.module_deadbeef.x86_64
+
         # The rpm-map exists to link checksums from repomd to specific
```
Add `# rpm-map:` here, and add other hard-coded keys throughout the rest of this block.

sgallagher commented on yaml_specs/modulemd_stream_v2.yaml:
```diff
@@ -134,14 +231,31 @@ data:
+        # Type: OPTIONAL
         tracker: http://www.example.com/
+
+    # profiles:
```
I'm going to defer to @contyk here. Petr, how much of this is accurate?

contyk replied:
Unfortunately, "default" has no special meaning nowadays. "buildroot" and "srpm-buildroot" are the only snowflakes we have. I'm not sure what youy mean by the component rpm flags.

sgallagher replied:
@contyk He's asking if we should include a pointer to `data.components.rpms.<name>.[srpm]buildroot` boolean attributes. I think the answer is yes.

mmathesius replied:
Yes, that's what I meant.

mmathesius commented on yaml_specs/modulemd_packager_v2.yaml:
```diff
@@ -3,15 +3,18 @@
 # for a module stream. It is always a proper subset of a `document: modulemd`
 # of the same `version`.
 document: modulemd-packager
```
One more comment... Once the `modulemd_stream_v2.yaml` text is settled, it would be best to tediously replicate the applicable pieces here in `modulemd_packager_v2.yaml` as well.

sgallagher replied:
Yeah, it probably would. :sob:

contyk commented on yaml_specs/modulemd_stream_v2.yaml:
```diff
@@ -1,50 +1,125 @@
+# the attribute and including it in the repository metadata. If the packager #
+# specifies this attribute explicitly, it must be preserved and provided in  #
+# the output metadata for inclusion into a repository.                       #
+##############################################################################
```
There are cases where the build system overrides supplied values; for instance, MBS has APIs that let you override the NSVC or the dependencies. I'm not sure if the language here isn't too strict.

While we normally want to it preserved, the abovementioned behavior shouldn't be considered broken or as violating the standard, I'd say.

sgallagher replied:
@contyk 
Does the following line sound right if added at the end of the glossary?

"The definitions above describe the expected behavior of the build system operating in its default configuration. It is permissible for the build system to override user-provided entries through non-default operating modes. If such changes are made, all items indicated as being required for the output repository must still be present."

contyk replied:
Yeah, that'd be fine.

sgallagher commented:
OK, I updated both the `document: modulemd` and `document: modulemd-packager` specifications to keep them in sync. I think I cleaned up the language of the `buildroot` options to make their purpose more clear.

In the process of reworking them to make them clearer, I had to update one of the tests which also revealed a minor bug in the output. We were emitting `buildorder` to the YAML as a 64-bit unsigned integer instead of as a 64-bit *signed* integer. This caused the test to fail when I set `buildorder: -1` in one place (since it was comparing `-1` to `MAX_UINT64`)

Another round of review would be helpful, please.

sgallagher commented:
@mmathesius Changes made

sgallagher commented:
Merged in 80b253eebee9caab151dc7e5b482b5fd89a32da1
